pkg/config/config.go

```diff
@@ -157,8 +165,8 @@ func (c *Config) LogDeprecations(l log.Logger) {
 	}
 }
 
-// ApplyDefaults sets default values in the config
-func (c *Config) ApplyDefaults() error {
+// Validate validates the config, flags, and sets default values.
+func (c *Config) Validate(fs *flag.FlagSet) error {
```

Comment:
I think this, alongside what #1142 (short term) and #1140 (longer term) may need, is perhaps calling for a better way of how we handle configs across the codebase. 

Today, we have many config structs spread across many packages. These config structs are (essentially) globals for a subsystem, where they contain their YAML schema, command line flags, and any other runtime value we needed to pass to them. Combining these like this complicates how we need to load config files, where the current process is as follows:

1. First, we define flags, which sets defaults against our struct. Flags take precedence over YAML, where a setting may be defined in both places (shoutout to my best friend #581).
2. Then, we parse flags just so we know what the config file flag is set to.
3. Next, we retrieve and unmarshal the config file as YAML. 
4. Finally, we parse the flags _again_, since the values set from the definition have almost definitely be overridden. Parsing again will give provided flags the proper precedence. 

Unmarshaling as YAML also tends to be a troublemaker because of how we attempt to provide defaults. We implement default settings through the following:

```go
var Default = Config{ ... } 
type Config struct { ... }

// UnmarshalYAML implements yaml.Umarshaler and sets c to Default 
// and unmarshals from YAML. Fields set in the YAML will override 
// the default values.
func (c *Config) UnmarshalYAML(f func(interface{}) error) error {
  *c = Default 
  type config Config 
  return f((*config)(c)) 
}
```

To make this even more complicated, #1142 will almost definitely require using a different schema for integrations if the integrations v2 feature is enabled. Maybe we can find a way to avoid this for now, but I generally don't think you can always avoid the need for runtime schema changes in a project where you do experimental features. #1140 is another good candidate for maybe having a different schema for the metrics subsystem if metrics v2 is enabled via flag. 

Even in the future with #581, we're still going to have this problem where loading configs is very convoluted and smelly. We may be able to separate out types in subsystems into dynamically configurable (YAML, `Config`) statically configurable (flags, `Options`). That might help with the complexity, but not much with the dynamic schema bit, and it all generally still smells a little stinky.

This is a lot of context with no answers, and I'll think on this more tomorrow. It feels like we have a few separate problems right now:

1. Flags and YAML are fighting with each other. We don't _want_ to support both. The YAML file will always need to be dynamically reloadable. Anything that is not dynamically reloadable should be a flag. 

2. Our configs struct are data monoliths that hold a cosmic horror of configs coming from a combination of YAML, flags, or generated at runtime after the config is loaded. 

3. There is the possibility of a near future where we will need to change the schema used for subsets of the config at runtime depending on the set of enabled feature flags. 

I'm hoping there's an obvious solution that I'm missing here that everyone else is already using :)

Reply:
> There is the possibility of a near future where we will need to change the schema used for subsets of the config at runtime depending on the set of enabled feature flags.

@sdboyer not to continue to complicate things, but...

Reply:
> We might be able to [...] decouple the config of a subsystem from how it is populated.

One final thought: if I were to rewrite everything from scratch, this is definitely the approach I would try taking first. I would [try] to treat our subsystems as libraries which are agnostic to how they are configured, and push the responsibility of how configs are represented to end users closer to the main package. 

Doing it that way would almost definitely give you a depressing amount of repetition when you inevitably define all your types twice to create a mapping of YAML -> agnostic config, but it would also make it a lot easier to be consistent and avoid the four problems we're struggling with now.

Reply:
>> There is the possibility of a near future where we will need to change the schema used for subsets of the config at runtime depending on the set of enabled feature flags.

> @sdboyer not to continue to complicate what you and I were talking about

i ain't scurred 😝 

Reading over this main system/subsystem distinction, it sounds like what you're aiming for is dynamic schema composition. That's actually quite similar to what Grafana needs - eventually, when a plugin is installed on a given Grafana instance, we'll need to be able to load up the CUE representing its scuemata lineage and inject it into the overall schema for dashboards. Accommodating different injection patterns is why we have three 'variants' of the dashboard: [base](https://github.com/grafana/grafana/blob/main/packages/grafana-schema/src/scuemata/dashboard/dashboard.cue#L385) (purely the root/system properties), [dist](https://github.com/grafana/grafana/blob/main/packages/grafana-schema/src/scuemata/dashboard/dist/family.cue) (base + all core plugins), instance (instance + plugins installed on a particular instance).

That composition system, specifically the "instance" case, seems like precisely what you're looking for here. And while the dist approach is now encoded directly in CUE, my original implementation was assembled dynamically in pure Go, and this [compose](https://github.com/grafana/grafana/blob/main/packages/grafana-schema/src/scuemata/dashboard/dashboard.cue#L385) pattern is designed specifically to be usable in either approach.

Reply:
Ah, lovely, scuemata to the rescue yet again :)

Reply:
> To make this even more complicated, #1142 will almost definitely require using a different schema for integrations if the integrations v2 feature is enabled. This would mean that Validate would need to exist purely for the purpose of reading flags so it can influence how UnmarshalYAML works. Maybe we can avoid needing this for #1142, but I generally think we'll struggle to avoid experimental schema changes. (#1140 seems likely to do it when the metrics v2 feature is enabled).
> 
> [...]
>
> I'm hoping there's an obvious solution that I'm missing here that everyone else is already using :)

🎉: Deferred/partial unmarshaling: https://github.com/grafana/agent/pull/1185